kernel/src/grant.rs

```diff
@@ -93,7 +93,9 @@ impl Allocator {
                             // We use `ptr::write` to avoid `Drop`ping the uninitialized memory in
                             // case `T` implements the `Drop` trait.
                             write(ptr, data);
-                            Ok(Owned::new(ptr, self.appid))
+                            // Unchecked is safe as we just created this
+                            let data = NonNull::new_unchecked(ptr);
```

Comment:
It seems weird to me that there's no direct way of creating a `NonNull<T>` from a `&T`.

Reply:
From the docs:
> Notice that NonNull<T> has a From instance for &T.

I think we could, but that's a `*T` not a `&T` anyways, so I don't think there's a safe conversion in either case.

Reply:
Somehow I had assumed that `arr` was a reference, but no it's actually a slice. However, I find it confusing that `process.alloc` returns a slice (building it via `unsafe { slice::from_raw_parts_mut(...) }`), and then we immediately convert it back to a pointer here. Is there a case where the result of `process.alloc` is used as a slice?

Otherwise, it would be simpler to have `process.alloc` return an `Option<NonNull<u8>>` and then here (in the `Some(non_null_ptr)` case) we can convert it to a `NonNull<T>` directly (via `.cast::<T>()`). This would avoid two unneeded `unsafe` operations: `slice::from_raw_parts_mut` and `NonNull::new_unchecked`.

Reply:
I looked at the Rust alloc interface, and it returns a pointer, so changing process.alloc() to return a pointer seems reasonable.

Reply:
From my searching, `process.alloc` is used only twice, and in both cases in the grant module -- I'll update this

Reply:
This again ended up refactoring a bit more, but I think the `alloc` inside of grant is much cleaner now as well. Resolved by dcb092e05f